### ISSUE TITLE
Work plan PMP 2017/2018

### ISSUE THREAD

**Issue by eguil** [CLIVAR-PRP/ENSO_metrics#4]
ENSO metrics notes 26/7/2017
Discussion:
- design modular system for ENSO metrics that allows separation of concern and governance by appropriate experts
- build on tests done at PCMDI on Bellenger et al. paper by Ji Woo and Peter

Ideas:
- list of CLIVAR ENSO approved command lines to run enso_bellenger_compute.py (https://github.com/PCMDI/pcmdi_metrics/tree/issue_321_enso_bellenger_2017/src/python/devel/enso_bellenger)
- develop driver_enso_metrics.py from enso_bellenger_compute.py API (Ji Woo to advise, Yann to test in ENSO metrics environment)
- further develop enso_metrics_lib.py on github from programs by Ji Woo et al., further documenting the science - Yann and interface with ENSO CLIVAR
- interface with CDAT/PMP libraries by Ji Woo and Peter
- interface between CLIVAR ENSO science and metrics lib by Yann and Eric
- provide reference json/graphics/table for past CMIPs : Ji Woo and Peter

Actions, timeline (and who):
- telco late August to firm up work plan and timeline (Yann to organise)
- produce first draft of work plan early Sept. (Eric to lead)
- review by CLIVAR RF mid-Sept. (Andrew and Eric to lead)
- implement first version by late Sept. (Yann and Ji Woo to lead)
- present at Busan CLIVAR ENSO RF mid Oct (Yann and Eric)
- review science contents and make final public release Nov 2017

**Comment by eguil** [comment]
Meeting Jiwwo/Yann/Eric on Aug 31st:

- Action Eric: access github x 3
- Update enso_metrics_lib.py with more metrics from enso_bellenger_compute.py (Action: Yann and Eric)
- Create a metrics collection (Action: Yann and Eric)
- Modify enso_bellenger_compute.py into a driver_enso_metrics.py (Action Jiwoo)
- Create a init_enso_metrics_obs.py (check if right data is there or download) (run by IT driver) 
- Args: MCi, local_obs_path
- Check against a enso_ref_obs.py/json listing var/obs_file_name/var_name (defined by ENSO experts)
- Action: produce draft enso_ref_obs.py (Yann & Eric)
- Action: produce init_enso_metrics_obs.py (Jiwoo)

**Comment by eguil** [comment]
Meeting Jiwoo/Peter/Yann/Eric on Sept. 14 st:

- action: Jiwoo to use new `initEnsoMetrics.py` and `metricsCollectionsLib.py` in driver all the way to run
- action: Yann to organise local set up using same driver
- action: Yann and Eric to add other metrics/collections
- action: review metrics calculation structure to include difference as well as model + obs calculation (Yann)

**Comment by eguil** [comment]
Meeting Jiwoo/Peter/Yann/Eric on Sept. 28th:
- organise exchange between PMP github and ENSO metrics gihub (PG & Jiwoo)
- develop proposed `ENSOCollectionsLib.py` and `ENSOMetricsLib.py` (Yann)
- `init.py` not needed (belong to driver)
- work on driver (Jiwoo and PG)

**Comment by eguil** [comment]
Meeting Jiwoo/Yann/Charles/Eric on Oct 12th:

- Charles describes new organisation on github towards creating conda packages (when the code is stabilized)
- For now Jiwoo will work in the `scripts/` directory (move `parser.py` back to `scripts/`)
- Jiwoo presents new devs on driver
- discussion how to define dates for observation and models (test on common/min length) - will live in `ENSOCollectionsLib.py`
- discussion on re-organising the metrics collection dicts (Yann and Eric)

**Comment by eguil** [comment]
Meeting Jiwoo/Yann/Peter/Eric on Oct 27th:

- Busan meeting summary 
[ENSO Research Foci Busan Oct 2017.pptx](https://github.com/eguil/ENSO_metrics/files/1420982/ENSO.Research.Foci.Busan.Oct.2017.pptx)
- IS-ENES3 discussions
- Discussion on display methods of MME results (Jiwoo / Peter)
- List of observations used for Obs4MIPs (Yann)
- rest of actions from last time
- next meeting Nov 2nd (Jiwoo and Yann)

**Comment by eguil** [comment]
Meeting Jiwoo/Yann/Peter/Eric on Nov 10th:



Ongoing actions:

- Document different display methods for MME results (Jiwoo / Peter)
 - example for ENSO metrics
- List of observations used for Obs4MIPs (Yann)

**Comment by eguil** [comment]
Meeting Jiwo,Yann, Eric on Nov 30th:
- Jiwoo updated driver to new organisation of metrics collection (and corrected some bugs)
- ENSO RF telco on ENSO metrics def last week - see channel #enso_rf_discussion in slack
- **TODO**: propose `compute_collection()` (which may include a loop on`compute_metric()`):
  - compute metric as model-obs distance for each metric for one model
  - have to define this distance calculation for each metric
  - the idea is to leave as little as possible to driver (only loop on models, path to files)
  - propose normalisation options to plot all metrics on same figure in a later stage

Ongoing actions:

- Document different display methods for MME results (Jiwoo / Peter) [example for ENSO metrics]
- List of observations used for Obs4MIPs (Yann)
- Regridding workflow

**Comment by eguil** [comment]
Meeting Jiwoo, Yann, Eric on Dec 14th:
- Yann demonstrates the new `compute_collection()` and associated procedures
- the term *metric* is now reserved for the distance between model and observation
- Driver will use info from `my_Param.py` to create dicts loop for `compute_collection()`
- new metrics to be implemented for next telco (TODO Yann)
- Jiwoo update use of database interface
- **TODO**: Jiwoo to adapt new version (commited now, comments needed)

Ongoing actions:

- Document different display methods for MME results (Jiwoo / Peter) [example for ENSO metrics]
- List of observations used for Obs4MIPs (Yann)
- Regridding workflow

**Comment by eguil** [comment]
Meeting Jiwoo, Yann, Eric on Dec 21st:
- Jiwoo's work last week:
  - confirmed the revised code by Yann which works well with the PMP driver
  - Build consistent XML files so I can easily apply all CMIP5 models to current version of code and parameter file
  - Revisited the portrait plot to clean up code and prepare documenting
- Yann's work:
  - created a branch for new version (called `yann` for now but we should find a new name)
  - further development of `compute_collection()` and associated tools (regrid, filter,...)
  - new MC2
- **TODO** [Jiwoo]: use `compute_collection()` from driver to remove loop on metrics - just keep loop on models 
- **TODO* [Eric]: start a shared document describing requirements for interface from this first experience


Ongoing actions:
- Document different display methods for MME results (Jiwoo / Peter) [example for ENSO metrics]
- List of observations used for Obs4MIPs (Yann)
- Regridding workflow

**Comment by eguil** [comment]
Meeting Jiwoo, Yann, Peter, Eric on Jan 10th:
- PG describes how an obs database is being developed in addition to obs4MIPs, using the same format (second bullet of ongoing actions)
- PG suggest to use several regridding to assess impact (ESMF, regrid2,...)
- EG suggest to not develop the interface in too much detail before we start interacting with ESMValTool.
- TODO [Jiwoo]: use `compute_collection()` from driver to remove loop on metrics - just keep loop on models

Ongoing actions:

- Document different display methods for MME results (Jiwoo / Peter) [example for ENSO metrics]
- List of observations used to be put in new ref obs database
- Regridding workflow

**Comment by eguil** [comment]
Meeting Jiwoo, Yann, Peter, Eric on Jan 26th:
- Yann and Peter: 2 normalization techniques illustrated in slack (vs. median or vs. ratio to obs)
  - may want to separate different statistics in multi window plots with different color scales
- PG: started to prepare observational database to follow obs4MIPs structure
- Jiwoo: driver now calls `compute_collection()` - see metrics_collection slack channel and `MC2` branch on github - required changes in obs database and internal organisation of driver
- Yann: developed MC2 collection (from first proposal by CLIVAR group) and tested them
- **TODO** [Yann]: explore how to override the collection default period per metric
- **TODO** [Jiwoo]: consolidate driver code, try different regrid
- **TODO** [Yann]: think about regrid specification (metric level, driver level, ?)
- first meeting with ESMValTool next week (Feb 1st)

Ongoing actions:
- TODO [Eric]: start a shared document describing requirements for interface from this first experience

**Comment by eguil** [comment]
Meeting Jiwoo, Yann, Peter, Eric on Jan 26th:
- Yann and Peter: 2 normalization techniques illustrated in slack (vs. median or vs. ratio to obs)
  - may want to separate different statistics in multi window plots with different color scales
- PG: started to prepare observational database to follow obs4MIPs structure
- Jiwoo: driver now calls `compute_collection()` - see metrics_collection slack channel and `MC2` branch on github - required changes in obs database and internal organisation of driver
- Yann: developed MC2 collection (from first proposal by CLIVAR group) and tested them
- **TODO** [Yann]: explore how to override the collection default period per metric
- **TODO** [Jiwoo]: consolidate driver code, try different regrid
- **TODO** [Yann]: think about regrid specification (metric level, driver level, ?)
- first meeting with ESMValTool next week (Feb 1st)

Ongoing actions:
- TODO [Eric]: start a shared document describing requirements for interface from this first experience

**Comment by eguil** [comment]
Meeting Jiwoo, Yann, Peter, Eric on Feb 2nd:
- summary of meeting with Veronika and Axel
- idea to move pre-preprocessing to driver using specific encoding in `ENSOCollectionsLibs.py`
- **TODO** [Yann+Jiwoo]: prepares a branch to test extracting pre-processing to driver
- **TODO** [Jiwoo]: explore how to put portrait plot and provenance on a web/wiki
- **TODO** [Jiwoo]: run MC2 to produce a portrait

Ongoing actions:
- Eric: start a shared document describing requirements for interface from this first experience (in wiki to start with)
- PG and JL to explore and document 1) regriding impact on results and 2) portrait plot normalization

**Comment by eguil** [comment]
Meeting Jiwoo, Yann, Peter, Eric on Feb 15th:
- Yann shows the pre-process branch with just one metric
- **TODO** [Jiwoo]: try again the MC2 branch to identify bug
- Once this is working, then move to pre-process branch and update the rest of MC1 and MC2
- Put pre-process branch back to master and work from there (create issues, related branches and pull back to master once issue is solved)

Ongoing actions:

- Eric: start a shared document describing requirements for interface from this first experience (in wiki to start with)
- PG and JL to explore and document 1) regriding impact on results and 2) portrait plot normalization

**Comment by eguil** [comment]
Meeting Jiwoo, Yann, Eric on Feb 22th
- 2 bugs identified (time checking, jason file issue wrt use of "mv2.multiply" vs. "*")
- jason file made for the IPSL model / MC1
- still an issue with one model
- Once this is working, then move to pre-process branch and update the rest of MC1 and MC2, put pre-process branch back to master and work from there (create issues, related branches and pull back to master once issue is solved)
- Jiwoo introduces Vega visualisation for jason files
- Meeting with ESMValTool group see #14 

Ongoing actions:

    - Eric: start a shared document describing requirements for interface from this first experience (in wiki to start with)
    - PG and JL to explore and document 1) regriding impact on results and 2) portrait plot normalization

**Comment by yyplanton** [comment]
Meeting (a quick hello from Eric) Jiwoo, Peter, Yann,  on Mar 8th

- uvcdat's cdutil.averager cannot handle irregular grids => we use areacello to average oceanic variables
- for now Jiwoo added areacello to the xml file of oceanic variables but it should be in separate files
- if we still have issues with averaging oceanic variables => use atmospheric ts (instead of tos) and tauu (instead of tauuo)

Ongoing actions:

- Yann: merge MC2 branch to master
- Yann: finalize preprocess branch
- Jiwoo: run preprocess branch and take ownership of pre-processing functions (Yann will help to understand and develop them)
- PG and JL to explore and document 1) regriding impact on results and 2) portrait plot normalization

**Comment by eguil** [comment]
Meeting Jiwoo, Peter, Yann, Eric Mar 20th
- MC1 and MC2 json file now working
- portrait plot in the making - issue of structure (nested dict + model AND obs)
  - leave models out that don't have the right data for now
- work on identifying ESMValTool pre-processing
- bring MC2 branch into master and create "ESMValTool" branch (evolution from preprocess)
- keep separate until ESMValTool interface works and then merge

Ongoing actions:
- PG and JL to explore and document 1) regriding impact on results and 2) portrait plot normalization 3) verification suite using correlations

**Comment by eguil** [comment]
Meeting Jiwoo, Peter, Yann, Eric on April 5th:
- MC2 merged in to master
- Portrait plots provided by Jiwoo
- relative errors shown wrt to mean error - ok to identify outliers
- *TODO* [Jiwoo] explore absolute error view:
  - parallel coordinates plot 
  - use of scale factor needed to show absolute error
  - add metric numbers in boxes to combine relative and absolute

Book keeping:
- use preprocess branch to investigate interface for ESMValTool
- keep separate until ESMValTool interface works and then merge in Master

Ongoing actions:

- PG and JL to explore and document 1) regriding impact on results and 2) portrait plot normalization 3) verification suite using correlations

**Comment by eguil** [comment]
Meeting Jiwoo, Peter, Yann, Eric on May 3rd:
- new portrait plots with actual metrics value as number on slack
- to start interacting with Andrew and al:
  - make version with atmos fields in new branch:atmos_regrid [Yann]
  - try other regridding options - add option in current code to allow overwrite by driver [Yann]
  [- add GFDL-CM2.1 (using atmos field is necessary) [Jiwoo]]
  - change relative into absolute metric calculation [Yann]
- think about adding other obs references (triangle)

Book keeping:
   - use preprocess branch to investigate interface for ESMValTool
   - keep separate until ESMValTool interface works and then merge in Master

Ongoing actions:
    - PG and JL to explore and document 1) regriding impact on results and 2) verification suite using correlations

**Comment by eguil** [comment]
Meeting Jiwoo, Peter, Yann, Eric on May 17th:
- work on regrid done by Yann ( to be commited soon) 
- move from matplotlib to VCS for graphics (better for interactive plots) done by Jiwoo
- add units [Jiwoo]
- create legend (online or doc) [Yann]
- rename MC2 into "ENSO perf"
- work on regrid (e.g. ts vs. tos): add option to have it per variable
- look into Yann being able to rerun the portrait plot by himself

Book keeping:
- use preprocess branch to investigate interface for ESMValTool
- keep separate until ESMValTool interface works and then merge in Master

Ongoing actions:

- PG and JL to explore and document 1) regriding impact on results and 2) verification suite using correlations
- think about adding other obs references (triangle)

**Comment by eguil** [comment]
Meeting Jiwoo, Peter, Yann, Eric on July 16th:
 -  json output reviewed to fit wider framework - update dict{} structure to fit this requirement (work around currently exists)
 - technical need to create a tag (e.g. v0.1) to create conda version to start integration in PMP
 - discuss science needs from telco with Andrew:
  - more portrait plot with absolute/relative metrics
    - interesting plot on slack (june 28th) mixing metric and relative MME color
    - TODO: [Jiwoo] have 2 portraits plots (RMSE & non RMSE)
  - // coord plots
  - dive down plots
  - science prioritisation (correlations between metrics)
 - new version of code pushed by Yann Friday
  - some dive down diags included (values used to compute the metric) 
  - [Jiwoo] to start producing a plot for dive down
 - MC news:
  - [Yann] start work on ENSO teleconnections



Ongoing actions:

 - PG and JL to explore and document 1) regriding impact on results and 2) verification suite using correlations
 - think about adding other obs references (triangle)
 - further organise provenance (both technical and scientific)

**Comment by yyplanton** [comment]
Meeting Jiwoo, Yann on July 26th:
- technical need to create a tag (e.g. v0.1): done. Jiwoo explained to Yann how to do it
- // coord plots created by Jiwoo (with or without lines): very nice to see the spread among models
   - it would be nice to be able to select several models for the plot and not always all of them (interactive plots)
   - is there a way to have the full markers for the first un last metrics of the plot? (and not half of it as it is now)
- discuss science needs from telco with Andrew:
   - more portrait plot with absolute/relative metrics
      - interesting plot on slack (june 28th) mixing metric and relative MME color
      - TODO: [Jiwoo] have 2 portraits plots (RMSE & non RMSE)
   - interesting // coord plots on slack (july 24th)
   - TODO: [Jiwoo] to start producing a plot for dive down (using the last version of the atmos_regrid branch)
   - science prioritisation (correlations between metrics)
 - MC teleconnections:
   - [Yann] first two metrics will be available quickly (still some science to discuss with Scott Power)
   - [Yann] these metrics will be used to test the json output reviewed to fit wider framework - update dict{} structure to fit this requirement 
- Ongoing actions:
   - PG and JL to explore and document 1) regriding impact on results and 2) verification suite using correlations
   - think about adding other obs references (triangle)
   - further organise provenance (both technical and scientific)

**Comment by eguil** [comment]
Meeting Jiwoo, Peter, Yann, Eric on Aug. 16th:
 - discussion on parallel plots:
  - very nice complement to portrait plot
  - issues with taux (2 models) and sst (?) 
 - version testing ready - will be included in PMP release for testing (hidden)
 - Yann: MC3 on ENSO teleconnection: branch Test_MC3
  - science discussion with Scott P on going
 - format of the dict for the json format
  - same depth level currently has mix of metrics (values) and meta-data (strings) 
  - fix structure to have only one type: how ?
   - work on package side or work on server side ?
   - Yann & Jiwoo to test: have 2 dict  (current and data only) embedded into one
  - Jiwoo and Peter to create an issue to document that, Yann uses the branch

- discuss science needs from telco with Andrew:
   - more portrait plot with absolute/relative metrics
      - TODO: [Jiwoo] have 2 portraits plots (RMSE & non RMSE)
   - interesting // coord plots on slack (july 24th)
   - TODO: [Jiwoo] to start producing a plot for dive down (using the last version of the atmos_regrid branch)
   - science prioritisation (correlations between metrics)
- MC teleconnections:
   - [Yann] first two metrics will be available quickly (still some science to discuss with Scott Power)
   - [Yann] these metrics will be used to test the json output reviewed to fit wider framework - update dict{} structure to fit this requirement 

Ongoing actions:
- PG and JL to explore and document 1) regriding impact on results and 2) verification suite using correlations
- think about adding other obs references (triangle)
- further organise provenance (both technical and scientific)

**Comment by eguil** [comment]
Meeting Jiwoo, Yann, Eric on Sept. 13th:
 - ENSO RF discussion two days ago
   - summary of progress presented - people are getting excited by the results
   - MC3 / teleconnection discussed

 - json format and dive down diags
   - branch to test ideas
   - pb with dive down diags (too long a file)
   - Start with option for dive down -> creates a new json file 
   - but netCDF will eventually be required
   - Yann will commit a version with new json structure (test by Jiwoo and bring back to trunk)

 - display topics
   - TODO: [Jiwoo] have 2 portraits plots (RMSE & non RMSE)
   - TODO: [jiwoo] // plot: issues with taux (2 models) and sst (clustering ?)
   - Jiwoo demonstrates an interactive // plot - link on slack


See above for other on-going actions

**Comment by eguil** [comment]
Meeting Jiwoo, Yann, Eric on Sept. 28th:
 - review tickets/issues
 - review branches
   - atmos_regrid can go
   - merge needed for json_structure - review pull request
 - correcting data
   - taux: reversed - correct data rather than the code (driver issue)
   - sst: using ocean-only points - ENSO metrics package calculation to integrate this
   - currently debugging on test_MC3 branch (use one/two models for now) 
 - Guayaquil material
   - 2 portrait plots
   - // coord. plot fixed
   - // coord. interactive

**Comment by eguil** [comment]
Meeting Jiwoo, Yann, Eric on Oct. 8th:
 - material for Guayaquil todo before end of week:
   - // plot fixed
   - portrait non-RMSE (MC1)
   - still an issue with some incoherent data/LS mask (driver issue ? e.g ACCESS)
   - taux issue:
     - fixed in data but not yet in portrait - either remove or fix
- Yann still investigating how issue was fixed (maybe interpolation and masking?) 
- branches reviews
   - Yann still investigating so merge will wait

**Comment by eguil** [comment]
Meeting Jiwoo, Yann, Peter, Eric on Nov. 15th:
- Meetings
  - ENSO Guayaquil - clear support
  - ENSO CLIVAR work
     - work on MC1 (5 new metrics) and MC2 (more changes coming up)
     - suggestions to add options in // plot - keep track of wish list in new issue
     - redo correlation in a more readable way and put on wiki
  - CMIP6 Barcelona abstract < 15 Dec
     - ENSO metrics
  - DOE meeting
     - overview of PMP and ENSO metrics - lot's of interest (other metrics such as MJO)
     - provide current work to demonstration site
- Dive down diags
    - stored in other json file - ok for 1D, more complex for 2D (use netCDF?)

**Comment by eguil** [comment]
Meeting Jiwoo, Yann, Peter, Eric on Nov. 28th:
 - option to save netCDF ready
 - ENSO CLIVAR work
   - work on MC1 (5 new metrics) and MC2 (more changes coming up) [TODO: Yann]
   - suggestions to add options in // plot - keep track of wish list in new issue [TODO:Yann]
   - [TODO:Yann]: redo correlation in a more readable way and put on wiki
   - add references to wiki metrics info
   - MC ENSO perform two portrait plots RMSE, non-RMSE - add description in metric collection file
 - CMIP6 Barcelona abstract < 15 Dec
   - ENSO metrics

**Comment by yyplanton** [comment]
Meeting Jiwoo, Yann on Dec. 19th:

    [TODO: Yann] option to save more information in netCDF
    ENSO CLIVAR work
        [TODO: Yann] work on ENSO_perf (one new metric has still to be coded: spectrum) and ENSO_tel (more changes coming up)
        suggestions to add options in // plot - keep track of wish list in new issue [TODO:Yann]
        [TODO:Yann]: redo inter-metrics correlation in a more readable way and put on wiki
        add references to wiki metrics info
        DONE: MC ENSO perform two portrait plots RMSE, non-RMSE - add description in metric collection file
       [TODO:Yann]: give Jiwoo me code to plot dive down diagnostics

**Comment by yyplanton** [comment]
Meeting Jiwoo, Yann on Dec. 19th:

    [TODO: Yann] option to save more information in netCDF
    ENSO CLIVAR work
        [TODO: Yann] work on ENSO_perf (one new metric has still to be coded: spectrum) and ENSO_tel (more changes coming up)
        suggestions to add options in // plot - keep track of wish list in new issue [TODO:Yann]
        [TODO:Yann]: redo inter-metrics correlation in a more readable way and put on wiki
        add references to wiki metrics info
        DONE: MC ENSO perform two portrait plots RMSE, non-RMSE - add description in metric collection file
       [TODO:Yann]: give Jiwoo me code to plot dive down diagnostics

**Comment by yyplanton** [comment]
Meeting Jiwoo, Yann on Dec. 19th:

  [TODO: Yann] option to save more information in netCDF
  ENSO CLIVAR work
        [TODO: Yann] work on ENSO_perf (one new metric has still to be coded: spectrum) and ENSO_tel (more changes coming up)
        suggestions to add options in // plot - keep track of wish list in new issue [TODO:Yann]
        [TODO:Yann]: redo inter-metrics correlation in a more readable way and put on wiki
        add references to wiki metrics info
        DONE: MC ENSO perform two portrait plots RMSE, non-RMSE - add description in metric collection file
       [TODO:Yann]: give Jiwoo me code to plot dive down diagnostics

**Comment by lee1043** [comment]
Closing this issue for house cleaning
